Fix song select realm refresh performance

peppy opened the pull request:
Closes https://github.com/ppy/osu/issues/25824

Each commit is basically standalone.

If any are seen as hard-to-merge or break test expectations, we can cherry-pick the good ones for now.

87b7699fcc29ab7e70ec1f632a0341bbe08a23b9 takes things from <1 FPS to a substantially usable frame rate with occasional stutters.

The rest of the commits work to reduce the stutter down to <100ms.

Can be tested using [test database](https://drive.google.com/drive/folders/1SMSR5v2YdbKXd78JiBHrxzKnex76Tgml?usp=sharing)

peppy committed:
Combine multiple similar invalidation logic into single event

peppy committed:
Avoid calling invalidation logic per beatmap set updated

Realm will batch the updates. We don't want to do expensive operations per set when we don't need to.

peppy committed:
Simplify invalidation logic

The only case where this was checking is guaranteed by realm to only be called once.

peppy committed:
Debounce count updates for good measure

peppy committed:
Cache non-filtered beatmap counts to massively improve count performance

peppy committed:
Improve performance of `attemptSelection` using new cached count and `LastSelected`

bdach commented on osu.Game/Screens/Select/SongSelect.cs:
```diff
@@ -162,7 +162,7 @@ private void load(AudioManager audio, OsuColour colours, ManageCollectionsDialog
                 SelectionChanged = updateSelectedBeatmap,
                 BeatmapSetsChanged = carouselBeatmapsLoaded,
-                FilterApplied = updateVisibleBeatmapCount,
+                FilterApplied = () => Scheduler.Add(updateVisibleBeatmapCount),
```
Why does this `Scheduler.Add()` while `carouselBeatmapsLoaded()` does `Scheduler.AddOnce()`?

peppy replied:
Just an oversight

bdach commented on osu.Game/Screens/Select/Carousel/CarouselGroup.cs:
```diff
@@ -55,6 +60,9 @@ public virtual void AddItem(CarouselItem i)
                 items.Add(i);
             }
+
+            if (!i.Filtered.Value)
+                TotalItemsNotFiltered--;
```
```suggestion
            if (!i.Filtered.Value)
                TotalItemsNotFiltered++;
```

surely?

bdach commented:
The invalidation logic changes are very difficult to follow. I'm still not sure whether behaviourally `BeatmapSetsLoaded` is always in the correct state now, or if `BeatmapSetsChanged` is invoked when it should be and not invoked when it shouldn't.

The test failures are not encouraging. Not sure if you want to look into them, or should I.

peppy committed:
Fix a few of silly issues

peppy committed:
Revert "Improve performance of `attemptSelection` using new cached count and `LastSelected`"

This reverts commit 25df42630ecba979457a9dcd2a7678d154a85a5e.

peppy commented:
I've reverted one change for now in the interest of getting this in. Will revisit it as a separate PR.

peppy commented:
One test remains flaky, trying to figure out what's going on but have not repro'd locally yet.

peppy committed:
Add realm refresh steps in an attempt to stabilise failing test

I think this is required because there is a higher chance of batched
updates with the new structure (and less calls to `BeatmapSetsChanged`
which causes re-selection).

bdach commented:
> Originally as far as I can tell it was only guarding in the realm callback, when the changeset is null. This is guaranteed by realm to only occur once, on initial event subscription.

No, `null` changeset will also happen after a realm refresh. Compare: https://github.com/ppy/osu/pull/25821.

I'm not sure what that means for this diff, since the rest of your answer is rather enigmatic, but there.

bdach committed:
Use nunit constraints in test for transparency

peppy committed:
Use `sender`'s realm (because we can)

peppy committed:
Actually fix realm selection retention regression

bdach commented:
Tests are failing because after beatmap removal the difficulty count under search bar does not update properly anymore. It shows stale counts from before the deletion.

Previously the deletion would happen before summing up the difficulty counts, but now it happens after. So this is a regression from https://github.com/ppy/osu/pull/25838/commits/8f5d21dc703e3c1f3930e4b4e1c58c90acaba87a specifically.

peppy committed:
Add missing invalidation call

peppy commented:
Easy fix, luckily. Just recalculate after deletion (should have been there anyway).

If another test breaks I will not be a happy peppy.

bdach commented:
Appears to be fine now, but I want this to go through a CI run with master merged before merging just to make sure I don't get clowned by unexpected crosstalk between this and the filtering fix.

bdach committed:
Merge branch 'master' into fix-song-select-realm-refresh-performance
